# COMS 4770 - Assignment 7 - Hedgren

In [1]:
import numpy as np

### 1. (a) Implement a procedure **evalPoly(n, a[], t, v[])** which takes as input the n + 1 coefficients of a polynomial of degree n stored in the array a[]. Here a[n] is the leading coefficient. At a point t, the procedure produces the values of the polynomial and its derivative in v[0] and v[1], respectively. Your procedure should support complex numbers.

In [10]:
"""
Evaluates a polynomial and its derivative at point t.

Parameters:
n (int): Degree of the polynomial.
a (list): Coefficients of the polynomial, where a[n] is the leading coefficient.
t (complex or float): The point where the polynomial and derivative are evaluated.
v (list): Output list where v[0] will store p(t) and v[1] will store p'(t).
"""
def evalPoly(n, a, t, v):
    poly_val = complex(0, 0)
    deriv_val = complex(0, 0)

    for i in range(n + 1):
        poly_val += a[i] * (t ** (n - i))
        if i < n:
            deriv_val += a[i] * (n - i) * (t ** (n - i - 1))

    v[0] = poly_val
    v[1] = deriv_val

#### (b) (6 pts) Use the procedure to evaluate the polynomial $p(x) = x^8 − 170x^6 + 7392x^4 − 39712x^2 + 51200$ at $1.414214$ and $1 + 2i$. Also, obtain its derivatives at these two points.

<span style="color:red">check output for t = 1+2j</span>

In [31]:
n = 8
a = [1, 0, -170, 0, 7392, 0, -39712, 0, 51200]

points = [1.414214, 1 + 2j]
results = []

for t in points:
    v = [0, 0]
    evalPoly(n, a, t, v)
    results.append((t, v[0], v[1]))

for t, poly_value, deriv_value in results:
    print(f"At t = {t}:")
    print(f"p({t}) = {poly_value}")
    print(f"p'({t}) = {deriv_value}")
    print()

At t = 1.414214:
p(1.414214) = (-0.015041687198390719+0j)
p'(1.414214) = (-34371.01227099437+0j)

At t = (1+2j):
p((1+2j)) = (98175-343400j)
p'((1+2j)) = (-446260-177000j)



##### Outputs:
At $t = 1.414214$:
$$p(1.414214) = (-0.015041687198390719+0j)$$
$$p'(1.414214) = (-34371.01227099437+0j)$$

At $t = (1+2j)$:
$$p((1+2j)) = (98175-343400j)$$
$$p'((1+2j)) = (-446260-177000j)$$

*NOTE: python's complex() uses j for denoting imaginary numbers instead of i*

### 2. In this problem you are asked to implement the fast Fourier transform and polynomial multiplication. Your code should have at least the following components:

#### (a) (11 pts) A procedure **DFT(n, a[], ahat[])** which takes as input the coefficients a[0], a[1], . . ., a[n] of a polynomial of degree n, where a[n] is the leading coefficient. It generates an array ahat[] that stores the images of these coefficients under the discrete Fourier transform.

<span style="color:red">check if omega_N is correct</span>

In [34]:
def DFT(n, a, ahat):
    omega_N = np.exp(2j * np.pi / n)
    
    for k in range(n):
        ahat_k = 0
        
        for j in range(n):
            ahat_k += a[j] * omega_N**(j * k)
        
        ahat[k] = ahat_k

n = 3
a = [1, 2, 3, 4]
ahat = [0] * (n)

DFT(n, a, ahat)

print("DFT coefficients (ahat):", ahat)

DFT coefficients (ahat): [(6+0j), (-1.5000000000000004-0.8660254037844377j), (-1.4999999999999984+0.8660254037844402j)]


#### (b) (6 pts) A procedure **IDFT(n, ahat[], a[])** which performs the inverse DFT. Namely, it outputs the coefficients a[0], a[1], . . ., a[n] from their images ahat[0], ahat[1], . . ., ahat[n] under DFT.

In [16]:
def IDFT(n, ahat, a):
    

#### (c) (5 pts) A procedure **multiplyPolys(n, m, a[], b[], c[])** which uses FFT to compute the product of two polynomials of degrees n and m, respectively. The coefficients of these two polynomials are stored in the arrays a[] and b[], respectively, with leading coefficients a[n] and b[m]. The coefficients of the product polynomial will be stored in the array c[].

#### (d) (4 pts) Use the procedure **multiplyPolys** to compute the product of the two polynomials below:

$p(x) = x^7 − 70.1x^6 + 2.4x^5 − 3.7x^4 + 7.4x^3 − 10.8x^2 + 10.8x − 6.8$

$q(x) = x^8 − 170x^6 + 0.614x^5 + 7392x^4 + 104.2x^3 − 39712x^2 + 51200$

##### Outputs: